### ISSUE TITLE
Constructor DSL confuses parameters of the same type

### ISSUE THREAD

**Issue by Nek-12** [InsertKoinIO/koin#1372]
When using two or more parameters of the same type when injecting, instead of passing `the value of the second argument` correctly to the expected parameter, the value of the other parameter is being passed.

**To Reproduce**
1. Create a class which has two parameters: `Int`, and `Int?` for example
```
class Foo(a: Int, b: Int?)
```
2. Inject the dependency 
```
val fooModule = module {
    factoryOf(::Foo)
}

// ... 

val foo by inject { parametersOf(5, null) } 
```
3. Observe the values
```
a = 5
b = 5 // unexpected
```

**Expected behavior**
```
a = 5
b = null
```

**Koin project used and used version (please complete the following information):**
`koin-core version 3.2.0`

**Comment by jonnyPap** [comment]
Hello 👋🏻 I face a similar issue. When declaring a class with 2 parameters of the same type (in my case `String`) without the need for one of them to be nullable.

Example:
```kotlin

class MyClass(
    val firstParam: String,
    val secondParam: AnotherClass,
    val thirdParam: String
): ViewModel()

viewModelOf(::MyClass)

val viewModel: MyClass by viewModels { parametersOf("hello", AnotherClass(), "there") }


//expected
viewModel.firstParam == "hello"
viewModel.thirdParam == "there"

//actual
viewModel.firstParam == "hello"
viewModel.thirdParam == "hello"

```

**Comment by andre-artus** [comment]
I ran into the same issue.

```
MyViewModel(
    private val param1: String,
    private val param2: String?,
    ...
```
param2 would be loaded with the value from param1.

**Comment by arnaudgiuliani** [comment]
If you have injection with similar types, I would advise to not use Constructor DSL as under the hood we fill with `get()`. we don't know what kind of params you need here

**Comment by Nek-12** [comment]
I think that's still a bug. I don't think anything is preventing the constructor DSL from iterating through the parameters normally, rather than just shoving the first one everywhere. Besides, it works with get() it seems but doesn't with the DSL

**Comment by arnaudgiuliani** [comment]
Perhaps more related to injected parameters. The thing is it will try to solve by type, not place index. Then your parameters are ambiguous, as you have 2 that have the same types. For such case, manual classic DSL is better.

But perhaps could help open an improvement topic regarding injected parameters 🤔

**Comment by gusandrianos** [comment]
Faced the same issue with 3 parameters of type `String` where all the parameters just had the same value of the first one.

**Comment by Nek-12** [comment]
The solution seems simple: inject resolved dependencies by type (in whatever way it works now), iterate through parameters sequentially + by filter by type.

**Comment by arnaudgiuliani** [comment]
the thing is, if you have 3 parameters of the same type. You propose to iterate one by one when pushing it in parameters?

**Comment by Nek-12** [comment]
Yes, my proposed solution that (I personally) would expect is that if I pass 3 parameters, they will preserve their order and amount. Please see my other issue where I've outlined all the current problems and a proposed solution

**Comment by arnaudgiuliani** [comment]
yes interesting. Let's keep the idea for a new patch in 3.4.0

**Comment by mvarnagiris** [comment]
Is there some workaround when using annotations? If I have @KoinViewModel that takes parameters of same type. How do I get instance of it with proper parameters?

**Comment by gusandrianos** [comment]
@mvarnagiris Better to use the old syntax for cases like these.

**Comment by mvarnagiris** [comment]
I ended up writing a wrapper object around the parameters I need. This solves the issue and I can still use annotations.

**Comment by bogdanzurac** [comment]
Was this closed because #1456 was opened instead?
